### ISSUE TITLE
Custom reporters produce no output if (1) markupsafe is imported before the run and (2) the target imports a missing module

### ISSUE THREAD

**Issue by b-kamphorst** [pylint-dev/pylint#8026]
### Bug description

It appears that (binary!) imports from `markupsafe._speedups` have side-effects in a `pylint` invocation. The file that is linted attempts to import a missing package in a `try-except` clause. A common invocation of `pylint` produces empty *stored* reports if `markupsafe` is imported *before pylint is linting*. However, output is produced as expected if either:
- the missing import is installed; or
- the `try` block is replaced by a non-import statement (e.g. `raise ValueError`); or
- the report is written to stdout instead of a file; or
- `markupsafe.__init__` is modified to prefer imports from `markupsafe._native`; or
- one imports a Cython package, e.g. `ormsgpack`, instead of `markupsafe`.

The first three observations may help in determining the root issue, but are not the problem themselves. The second two observations are what made me report the issue at `MarkupSafe`, as that suggests that the issue is specific to this package. However: *I don't know where the root cause is, so I'm submitting this issue to both MarkupSafe and Pylint and trust that the community has a more complete understanding of the powers at play.*

The issue at `MarkupSafe` (copy of this issue): https://github.com/pallets/markupsafe/issues/344.

This issue also lies at the root of https://gitlab.com/smueller18/pylint-gitlab/-/issues/18, where `MarkupSafe` is installed as a dependency of `jinja2`.

#### Setup

I have two files in the same directory: `invoke_pytest.py` and `my_module.py`
```py
# invoke_pytest.py

import markupsafe  # this import is not used -- it can only contribute side-effects or invoke pylint side-effects!
from pylint import run_pylint

run_pylint(
    argv=[
        "--output-format=text,text:report.txt",
        "my_module.py",
    ]
)
```
```py
# my_module.py

try:
    import click
except ImportError:
    pass
```

I have a virtual environment in which I installed `pylint` and `markupsafe`.

I understand that the example is not quite minimal yet -- there is a lot of pylint stuff under the hood. Again, let me emphasize that I reach out to both communities and I hope to jointly reduce the example shortly.

### Command used

```shell
`python invoke_pytest.py; echo "---- contents of report.txt ----"; cat report.txt`
```

### Pylint output

```shell
************* Module my_module
my_module.py:1:0: C0114: Missing module docstring (missing-module-docstring)
my_module.py:2:4: W0611: Unused import click (unused-import)

------------------------------------------------------------------
Your code has been rated at 5.00/10 (previous run: 5.00/10, +0.00)

---- contents of report.txt ----
```
So, we get output from `--output-format=text`, but `--output-format=text:report.txt` results in an empty file.

### Expected behavior

I would expect the output of pylint to stdout to also show up in `report.txt`. After commenting `import markupsafe` from `invoke_pytest.py`, the output is as expected:
```shell
************* Module my_module
my_module.py:1:0: C0114: Missing module docstring (missing-module-docstring)
my_module.py:2:4: W0611: Unused import click (unused-import)

------------------------------------------------------------------
Your code has been rated at 5.00/10 (previous run: 5.00/10, +0.00)

---- contents of report.txt ----
************* Module my_module
my_module.py:1:0: C0114: Missing module docstring (missing-module-docstring)
my_module.py:2:4: W0611: Unused import click (unused-import)

------------------------------------------------------------------
Your code has been rated at 5.00/10 (previous run: 5.00/10, +0.00)
```

**As mentioned above, the same output is achieved if (1) `markupsafe.__init__` is adjusted to not import from `._speedups`.** `_speedups` is written in C or (2) `click` is installed.

### Pylint version

```shell
pylint 2.17.0 (before: 2.15.9)
astroid 2.15.0 (before: 2.12.13)
Python 3.10.9 (main, Jan  3 2023, 17:26:40) [GCC 9.4.0]
```

### OS / Environment

Windows 10 -> WSL2 -> Ubuntu

### Additional dependencies

The most notable dependency below is `MarkupSafe==2.1.2` (before: 2.1.1).

```bash
$ pip list
Package           Version
----------------- -------
astroid           2.15.0
dill              0.3.6
isort             5.12.0
lazy-object-proxy 1.9.0
MarkupSafe        2.1.2
mccabe            0.7.0
pip               23.0.1
platformdirs      3.1.1
pylint            2.17.0
setuptools        65.5.0
tomli             2.0.1
tomlkit           0.11.6
typing_extensions 4.5.0
wrapt             1.15.0
```

I want to finish this report by stating that I appreciate all your efforts, keep up the good work!

**Comment by DanielNoord** [comment]
I don't really know where to start debugging this. Does this also happen without using a plugin and using a builtin output format?

**Comment by b-kamphorst** [comment]
> I don't really know where to start debugging this.

That I can imagine -- I already spent a fair amount of time getting this far and at this point there is too much stuff that I am not familiar with (Python + C, pylint detection and "import" of modules, ...).

> Does this also happen without using a plugin and using a builtin output format?

It does. I'll update the description to make the example more minimal.

**Comment by Pierre-Sassoulas** [comment]
Fixed by [pylint@98eb21b13](https://github.com/pylint-dev/pylint/commit/98eb21b13082d30bf72fcddd5b740e9d621c7149)
